kong/plugins/ldap-auth/access.lua

```diff
@@ -50,17 +50,28 @@ local function ldap_authenticate(given_username, given_password, conf)
 
   sock:settimeout(conf.timeout)
 
-  ok, err = sock:connect(conf.ldap_host, conf.ldap_port)
+  local opts = {}
+
+  -- keep TLS connections in a separate pool to avoid reusing non-secure
+  -- connections and vica versa, because StartTLS use the same port 
+  if conf.start_tls then
+    opts.pool = conf.ldap_host .. ":" .. conf.ldap_port .. ":starttls"
```

Comment:
Is there a reason why we would only want to set this pool option for TLS connections only? Is there a reason why we cannot use the default `pool` key generated by ngx_lua?

Reply:
The reason is that the default `pool` key only has IP and ports and since `STARTTLS` uses the same IP and port with cleartext connections, mixing them will create issues in reused connections (cleartext connection returned for `STARTTLS` request and already `STARTTLS`'d connections returned for cleartext connections).

Reply:
The reason for pool separation is that STARTTLS use non-secure port to initialize connections and connection pool name is generated by default in [HOST:PORT] pattern so it might be the case where non-secure and TLS connections be mixed up in the same pool, so thought it is a good thing to separate this particular pool with [HOST:PORT:starttls] name pattern.

Reply:
> same IP and port with cleartext connections

Ugh... Ok, perfect :+1: